### PR TITLE
DSDF-1 Channel Dropout: Force Shape-Invariant Tandem Prediction

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model currently sees all DSDF channels for both foils during every training step. For tandem samples, foil-1 DSDF channels (x[:,2:6]) encode the exact shape of the fore-foil — giving the model a shortcut: memorize fore-foil shape → predict downstream pressure. This shortcut works well for seen geometries (NACA0012-series) but degrades for unseen ones (NACA6416 in val_tandem_transfer).

**The fix:** During training, randomly zero out the foil-1 DSDF channels (x[:,2:6]) with probability p (applied per-sample, tandem samples only). This forces the model to predict aft-foil pressure even when fore-foil shape info is partially absent — building shape-invariant wake representations grounded in flow physics rather than geometry memorization.

This is inspired by:
1. The success of `--aug_dsdf2_sigma 0.05` (foil-2 DSDF magnitude augmentation, PR #2126) which improved p_tan by corrupting shape info
2. Dropout's proven effectiveness as a regularizer for forcing distributed representations
3. The NACA6416 representation gap: p_tan is the hardest metric because the aft-foil's unseen shape interacts with an equally-unseen fore-foil combination

**Key difference from aug_dsdf2_sigma:** σ scaling corrupts magnitude but preserves direction; channel dropout forces complete zero-shot shape generalization for random training steps. The model must predict aft-foil pressure with zero fore-foil shape information for those samples.

**Expected impact:** -2% to -5% p_tan. Single-foil samples completely unaffected (dropout is tandem-only).

### Step 1 — Add config flag

In `Config` dataclass:
```python
dsdf1_channel_dropout: float = 0.0   # p(zero out foil-1 DSDF channels) per tandem sample; 0 = off
```

### Step 2 — Verify foil-1 DSDF channel indices

Look at the dataset/feature construction code to confirm which feature indices correspond to foil-1 DSDF. Based on PR #2126 (foil-2 DSDF at x[:,6:10]), foil-1 DSDF is likely x[:,2:6]. Confirm before hardcoding.

### Step 3 — Apply dropout in training loop

In the training loop, after the batch is loaded and **before** passing to the model, apply the dropout. Add a `if model.training:` guard so validation is clean:

```python
if cfg.dsdf1_channel_dropout > 0 and model.training:
    # is_tandem: [B] bool tensor (already computed in batch loading)
    drop_mask = torch.rand(x.shape[0], device=x.device) < cfg.dsdf1_channel_dropout
    tandem_and_drop = is_tandem.bool() & drop_mask
    if tandem_and_drop.any():
        x = x.clone()  # avoid in-place modification of the original tensor
        x[tandem_and_drop, 2:6] = 0.0  # adjust indices if needed
```

### Step 4 — Run 3 experiments

Use `--wandb_group phase6/dsdf1-channel-dropout`.

| GPU | Config | wandb_name | Seed |
|-----|--------|-----------|------|
| 0 | p=0.2 | tanjiro/dsdf1-drop-p02-s42 | 42 |
| 1 | p=0.2 | tanjiro/dsdf1-drop-p02-s73 | 73 |
| 2 | p=0.3 | tanjiro/dsdf1-drop-p03-s42 | 42 |

Full command (p=0.2, seed 42):
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro \
  --wandb_name "tanjiro/dsdf1-drop-p02-s42" --wandb_group phase6/dsdf1-channel-dropout \
  --dsdf1_channel_dropout 0.2 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

Repeat with `--seed 73` and `--dsdf1_channel_dropout 0.3 --seed 42`.

**Do NOT override SENPAI_MAX_EPOCHS or SENPAI_TIMEOUT_MINUTES.**

## What to Report

- Surface MAE for all runs: p_in, p_oodc, p_tan, p_re with W&B run IDs
- 2-seed avg for p=0.2 vs baseline
- Does p_in regress? (Should not — single-foil samples unaffected)
- Which foil-1 DSDF indices did you use? (confirm from code inspection)

## Baseline

Current single-model baseline (PR #2130, GSB + PCGrad, 2-seed avg):

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in   | 13.05     | < 13.05        |
| p_oodc | 7.70      | < 7.70         |
| **p_tan** | **28.60** | **< 28.60** |
| p_re   | 6.55      | < 6.55         |

W&B baseline runs: `d7l91p0x` (seed 42, p_tan=28.9), `j9btfx09` (seed 73, p_tan=28.3)

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-gsb-pcgrad" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

All 3 experiments completed (180-181 min each, timeout-limited at ~151-152 epochs).

### Per-run results

| Run | Epoch | val_loss | p_in | p_tan | p_oodc | p_re | W&B |
|-----|-------|----------|------|-------|--------|------|-----|
| p=0.2, s42 | 151 | 0.3893 | 12.8 | 30.3 | 7.9 | 6.0 | vtma4eyd |
| p=0.2, s73 | 151 | 0.3884 | 13.1 | 30.1 | 7.3 | 5.9 | 8h5vj8of |
| p=0.3, s42 | 152 | 0.3906 | 12.3 | 30.4 | 7.4 | 6.0 | zwso4v4q |

### 2-seed averages (p=0.2) vs baseline

| Config | p_in | p_tan | p_oodc | p_re | Δ p_tan |
|--------|------|-------|--------|------|---------|
| **Baseline (no dropout)** | **13.05** | **28.60** | **7.70** | **6.55** | — |
| p=0.2 (2-seed avg) | 12.95 | 30.20 | 7.60 | 5.95 | +5.6% worse |
| p=0.3 (1 seed) | 12.3 | 30.4 | 7.4 | 6.0 | +6.3% worse |

### Foil-1 DSDF channel indices

Confirmed from code inspection: foil-1 shape info = x[:,:,2:6] (saf(2) + dsdf[0:2]), consistent with foil-2 at x[:,:,6:10] used by existing `aug_dsdf2_sigma`.

### Does p_in regress?

**No.** p_in actually improved: 12.95 vs 13.05 baseline (-0.8%). Single-foil samples are unaffected by the dropout (tandem-only guard), and in-distribution performance benefits slightly from the regularization.

### Analysis

**Hypothesis rejected for p_tan.** Both dropout rates made p_tan significantly worse (+5.6% for p=0.2, +6.3% for p=0.3). The model needs foil-1 shape information to predict tandem transfer correctly — zeroing it out removes critical wake interaction signal.

**However, notable improvements elsewhere:**
- p_re: 5.95 vs 6.55 (-9.2% improvement) — the dropout acts as strong regularization for OOD Reynolds
- p_oodc: 7.60 vs 7.70 (-1.3% improvement) — modest OOD condition gain
- p_in: 12.95 vs 13.05 (-0.8% improvement) — in-dist also slightly better

**Why p_tan got worse:** The foil-1 shape channels encode fore-foil geometry that directly determines the wake impinging on the aft foil. Zeroing these channels forces the model to predict aft-foil pressure "blind" to the upstream shape — which is fundamentally harder than the DSDF2 magnitude scaling from PR #2126 (which preserves gradient directions while only corrupting magnitudes). Channel dropout is too aggressive for the tandem transfer task.

**Key insight:** The strong p_re improvement suggests the dropout prevents overfitting to foil-shape-specific patterns. A softer version (e.g., scaling foil-1 DSDF by a random factor like aug_dsdf2_sigma does, rather than zeroing) might capture the regularization benefit without destroying critical wake geometry signal.

### Suggested follow-ups

- **Foil-1 DSDF magnitude augmentation** (like aug_dsdf2_sigma but for channels 2:6) — preserves gradient direction while corrupting magnitude. This was the winning approach for foil-2 and may work for foil-1 too.
- The strong p_re improvement from channel dropout could be combined with a domain-specific application: dropout only during OOD Reynolds training steps, or with lower probability (p=0.05-0.1).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2156 because ADVISOR: Closing — both dropout rates regress p_tan substantially.

**p=0.2:** avg p_tan=30.20 (+5.6% vs 28.60 baseline)
**p=0.3:** avg p_tan=30.40 (+6.3% vs baseline)

Zeroing foil-1 DSDF channels removes the critical wake-interaction signal that the model needs for tandem prediction. The foil-1 DSDF encodes the upstream geometry that creates the wake impinging on the aft foil — dropping it forces the model to predict without this causal information.

Interesting finding: p_re improved -9.2% at p=0.2, suggesting the dropout acts as strong regularization for OOD Reynolds generalization. But the p_tan regression is too severe to justify.

Note: foil-1 DSDF magnitude augmentation (like aug_dsdf2_sigma for channels 2:6) was already tested in PR #2133 and all σ values regressed p_tan. The foil-1 DSDF channels appear to need exact values — any perturbation (dropout or magnitude noise) hurts tandem transfer.

Dead end: foil-1 DSDF channel perturbation in any form (dropout, magnitude aug, mixup). These channels carry precise upstream geometry signal.

Thanks tanjiro — good exploratory work and the p_re finding is noted..
